axlearn/common/logging_utils.py

```diff
@@ -0,0 +1,138 @@
+# Copyright © 2025 Apple Inc.
+
+"""Logging utility classes"""
+
+import logging
+import threading
+import time
+
+
+class LoggingContext:
+  """A context manager for monitoring the execution time of operations and
+  logging warnings based on defined timeouts and thresholds.
+
+  This context manager can be used to:
+  - Log a warning if an operation exceeds a specified timeout while still
+  running.
+  - Log a warning if an operation's total execution time exceeds a specified
+  threshold.
```

Comment:
It's not clear that we need logging-upon-timeout, given the threading concerns raised by @Ethanlm. I wonder if a simpler semantic would be sufficient:

- LoggingContext always wait until the op finishes and log a warning if the execution time exceeds the threshold;
- We rely on the watchdog https://github.com/apple/axlearn/blob/2c4ec6ea939a12b6752ae3f95fbc4fcc58388dad/axlearn/common/trainer.py#L190 to tell us where the threads are stuck

WDYT?

Reply:
I agree with starting something simpler so I changed this PR to only implement logging upon op completion. 

However, I feel logging-upon-timeout is important because program hangs are difficult to debug by just looking through tracebacks of so many workers printed by watchdog timeout. A clear logging message may help identify workers that cause the hangs.